src/data/bucket/circle_bucket.js

```diff
@@ -86,25 +87,32 @@ class CircleBucket<Layer: CircleStyleLayer | HeatmapStyleLayer> implements Bucke
         }
 
         for (const {feature, id, index, sourceLayerIndex} of features) {
-            if (this.layers[0]._featureFilter(new EvaluationParameters(this.zoom), feature)) {
-                const geometry = loadGeometry(feature);
-                const sortKey = circleSortKey ?
-                    circleSortKey.evaluate(feature, {}) :
-                    undefined;
-
-                const bucketFeature: BucketFeature = {
-                    id,
-                    properties: feature.properties,
-                    type: feature.type,
-                    sourceLayerIndex,
-                    index,
-                    geometry,
-                    patterns: {},
-                    sortKey
-                };
-
-                bucketFeatures.push(bucketFeature);
-            }
+            const needGeometry = this.layers[0]._featureFilter.needGeometry;
+            const evaluationFeature = {type: feature.type,
+                id: ('id' in feature ? feature.id : null),
```

Comment:
You need to use `id` rather than `feature.id` here for the GeoJSON `promoteFeatureID` feature to work.
```suggestion
                id,
```

Reply:
I tried with this way, but it will cause the render test `regressions/mapbox-gl-js#4235` failure, as id here is `undefined` instead of `null`,  as `filter-has-id` only compare with `null` here
https://github.com/mapbox/mapbox-gl-js/blob/e5db688fe1a583aa002a5387b15218d6a9d3f949/src/style-spec/expression/definitions/index.js#L455-L458

Reply:
Maybe we should change this expression to `!=` (which also compares against `undefined`) then?

Reply:
Agree! I was afraid to do so because I wander if there is some special logic behind it. I will change it!